### PR TITLE
Implementation of fast fourier transform for discrete signals in C

### PR DESCRIPTION
**Pull request by pri2810**
Adding code for a fast_fourier_transform  category of mathematical_algorithms in C language.

### CODE REVIEW

File: code/mathematical_algorithms/src/fast_fourier_transform/fft.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
+/* Part of Cosmos by OpenGenus Foundation */
```
Comment (abdouskamel) [inline code/mathematical_algorithms/src/fast_fourier_transform/fft.c]: Be sure to look at the C style guide.

**Comment by pri2810** [comment]
I went through the C style guide and made the necessary changes. Please guide me further.

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 2]
```diff
@@ -0,0 +1,121 @@
+#include<stdio.h>
```
Comment (abdouskamel) [inline code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Spaces after include please.

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+twiddle( int n, double w[] )
+{
+	double arg;
+	double aw;
+	int i;
+	const double pi = constant/2;
+
+	aw = 2.0 * pi / ((double)n);
+
+	for ( i=0;i<n/2;i++ )  {
```
Comment (abdouskamel) [inline code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Spaces after ;

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		arg = aw * ((double)i);
+		w[i*2+0] = cos ( arg );
```
Comment (abdouskamel) [inline code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Spaces after arithmetic operators.

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	return;
+}
+
+/* FFT */
+void
+FFT(double *x, double *y, int n, double w[])
+{
+
+	int m,i,j,k,nbits;
```
Comment (abdouskamel) [inline code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Spaces after ,

Comment (abdouskamel) [reply code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Same for all your file, add spaces after operators, semicolons ...etc.

**Comment by abdouskamel** [comment]
@pri2810 I can just suggest you style improvement. For the algorithm itself, I can't review it because I do not master it. I hope that someone will be able to review it and tells if it's correct or not. Otherwise, you should test it in multiple and heterogeneous inputs.

**Comment by pri2810** [comment]
Thank you @abdouskamel for suggesting the style improvement. From my side, I have already tested it for various inputs and confirmed them with the Discrete transform outputs. Also, is it necessary to have user input or just a hardcoded input?

**Comment by abdouskamel** [comment]
@pri2810 It's ok for user input. :+1:

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 6]
```diff
@@ -0,0 +1,121 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#define constant 6.28318530718
```
[the remark this answers is not in the document]
Comment (pri2810) [reply code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Thanks for the information

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+/* look up table */
+void 
+twiddle( int n, double w[] )
```
[the remark this answers is not in the document]
Comment (pri2810) [reply code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Ohh! That's a small thing that was left out! Thank you for mentioning

File: code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+{
+	double arg;
+	double aw;
+	int i;
+	const double pi = constant / 2;
+
+	aw = 2.0 * pi / ( ( double ) n );
```
[the remark this answers is not in the document]
Comment (pri2810) [reply code/mathematical_algorithms/src/fast_fourier_transform/fast_fourier_transform.c]: Thank you @arnavb  for enlighting further on the style guide of writing a C code. Please inform me if there are any other major guidelines that I may have missed

**Comment by pri2810** [comment]
Thank You, @arnavb.

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**